### PR TITLE
fix: correct types for some customization options

### PR DESCRIPTION
**Pull request by lebensterben**

### CODE REVIEW

File: elpaca.el [hunk 1]
```diff
@@ -67,7 +67,8 @@
                                  (failed   . elpaca-failed)
                                  (busy     . elpaca-busy))
   "Alist mapping order statuses to faces."
-  :type 'alist)
+  :type '(alist :key-type (:options (blocked finished failed busy))
+                :value-type face))
```
Comment (progfolio) [inline elpaca.el:71]: Unfortunatley this throws an error for me on Emacs 30 and 27:

0. apply patch
1. `M-x customize-group` `elpaca`
2.  Click the "Show value" button (or arrow depending on Emacs version) to the left of Elpaca Status Faces in the Customize Group: Elpaca buffer.

Expected behavior: The value options are shown.
Actual behavior: An error is signaled (see below) and the widget is removed from the buffer.

<details><summary>error backtrace</summary>

```emacs-lisp
Debugger entered--Lisp error: (void-function nil)
  nil((:options :args ((blocked finished failed busy))) blocked)
  widget-match-inline((:options :args ((blocked finished failed busy))) (blocked elpaca-blocked))
  widget-group-match-inline((cons :format "%v" :args ((:options :args ((blocked finished failed busy))) (face :value "default"))) (blocked elpaca-blocked))
  widget-group-match((cons :format "%v" :args ((:options :args ((blocked finished failed busy))) (face :value "default"))) (blocked elpaca-blocked))
  widget-cons-match((cons :format "%v" :args ((:options :args ((blocked finished failed busy))) (face :value "default"))) (blocked . elpaca-blocked))
  widget-match-inline((cons :format "%v" :args ((:options :args ((blocked finished failed busy))) (face :value "default"))) ((blocked . elpaca-blocked) (finished . elpaca-finished) (failed . elpaca-failed) (busy . elpaca-busy)))
  widget-editable-list-match-inline((editable-list :inline t :args ((cons :format "%v" :args ((:options :args ((blocked finished failed busy))) (face :value "default"))))) ((blocked . elpaca-blocked) (finished . elpaca-finished) (failed . elpaca-failed) (busy . elpaca-busy)))
  widget-match-inline((editable-list :inline t :args ((cons :format "%v" :args ((:options :args ((blocked finished failed busy))) (face :value "default"))))) ((blocked . elpaca-blocked) (finished . elpaca-finished) (failed . elpaca-failed) (busy . elpaca-busy)))
  widget-group-match-inline((alist :key-type (:options (blocked finished failed busy)) :value-type face :args ((editable-list :inline t :args ((cons :format "%v" :args ((:options :args ...) (face :value "default"))))))) ((blocked . elpaca-blocked) (finished . elpaca-finished) (failed . elpaca-failed) (busy . elpaca-busy)))
  widget-group-match((alist :key-type (:options (blocked finished failed busy)) :value-type face :args ((editable-list :inline t :args ((cons :format "%v" :args ((:options :args ...) (face :value "default"))))))) ((blocked . elpaca-blocked) (finished . elpaca-finished) (failed . elpaca-failed) (busy . elpaca-busy)))
  custom-variable-value-create((custom-variable :group ... :tag ... :custom-prefixes ... :custom-level 2 :value elpaca-status-faces :parent ... :indent ...))
  widget-default-create((custom-variable :group ... :tag ... :custom-prefixes ... :custom-level 2 :value elpaca-status-faces :parent ... :indent ...))
  widget-default-value-set((custom-variable :group ... :tag ... :custom-prefixes ... :custom-level 2 :value elpaca-status-faces :parent ... :indent ...) elpaca-status-faces)
  widget-value-set((custom-variable :group ... :tag ... :custom-prefixes ... :custom-level 2 :value elpaca-status-faces :parent ... :indent ...) elpaca-status-faces)
  custom-redraw((custom-variable :group ... :tag ... :custom-prefixes ... :custom-level 2 :value elpaca-status-faces :parent ... :indent ...))
  custom-toggle-hide-variable((custom-visibility :args nil :value nil :help-echo "Show the value..." :on-glyph "down" :on "Hide" :off-glyph "right" :off ...) nil)
  widget-apply-action((custom-visibility :args nil :value nil :help-echo "Show the value..." :on-glyph "down" :on "Hide" :off-glyph "right" :off ...) nil)
  Custom-newline(2283)
  funcall-interactively(Custom-newline 2283)
  command-execute(Custom-newline)
```


</details>


Does this type spec allow for adding arbitrary `(status . face)` cells in addition to the listed `:options`? Users (and third party packages) may want to augment the list with their own options.

File: elpaca.el [hunk 2]
```diff
@@ -107,10 +108,10 @@ Results in faster start-up time." :type 'boolean)
   "Location of the repos directory.")
 
 (defcustom elpaca-makeinfo-executable (executable-find "makeinfo")
-  "Path of the makeinfo executable." :type 'string)
+  "Path of the makeinfo executable." :type '(file :must-match t))
 
 (defcustom elpaca-install-info-executable (executable-find "install-info")
-  "Path of the install-info executable." :type 'string)
+  "Path of the install-info executable." :type '(file :must-match))
```
Comment (progfolio) [inline elpaca.el:114]: Should there be a `t` value for `:must-match` here as well?
That is the only value I see documented in the manual.

**Review by progfolio — CHANGES_REQUESTED** [review]

File: elpaca.el (hunk 1 above)
Comment (lebensterben) [reply elpaca.el:71]: Unfortunately I cannot reproduce this error on 28.2.

2023-04-19-002443_2560x1440_scrot

Comment (progfolio) [reply elpaca.el:71]: I'm able to reproduce on 28.2. I tested on 27 and 28 via a docker image w terminal Emacs and
30 on my personal machine (linux) with GUI and terminal.
Same issue on all three for me. I'd say it'd be better to exclude this change for now until someone can figure out how to make it work for all of those.
I'd be happy to accept the other changes.

Comment (lebensterben) [reply elpaca.el]: @progfolio 

After `elpaca-rebuild`, I'm able to reproduce this error.
I've pushed the fix.

File: elpaca.el [hunk 3]
```diff
@@ -67,7 +67,8 @@
                                  (failed   . elpaca-failed)
                                  (busy     . elpaca-busy))
   "Alist mapping order statuses to faces."
-  :type 'alist)
+  :type '(alist :key-type symbol (:options (blocked finished failed busy))
+                :value-type face))
```
Comment (progfolio) [inline elpaca.el:71]:
```suggestion
  :type '(alist :key-type symbol :options '(blocked finished failed busy) :value-type face))
```

From the elisp manual section on "Composite Types":

>     The argument to the ‘:options’ keywords should be a list of
     specifications for reasonable keys in the alist.  Ordinarily, they
     are simply atoms, which stand for themselves.  For example:

          :options '("foo" "bar" "baz")

That should provide properly labeled checkbox widgets:
image

**Comment by lebensterben** [comment]
@progfolio I committed your suggestion.

File: elpaca.el [hunk 4]
```diff
@@ -880,7 +881,8 @@ If it matches, the E associated with process has its STATUS updated."
       (setq lines (butlast lines)))
     (dolist (line lines)
       (unless (string-empty-p line)
-        (elpaca--signal e (car (last (split-string line "" t)))
+        (elpaca--signal e (car (last (split-string line "
+" t)))
```
Comment (progfolio) [inline elpaca.el:885]: Unrelated changes?

Comment (progfolio) [inline elpaca.el:885]: Here, too

Comment (lebensterben) [reply elpaca.el:885]: That's strange. It doesn't show up on my local. And 

git diff [f53b8c](https://github.com/lebensterben/elpaca/commit/f53b8c25533b79878cca20420e29a03163b2ed9b)

shows nothing

Comment (progfolio) [reply elpaca.el]: Interesting. Git configured to ignore white space changes maybe?

**Comment by lebensterben** [comment]
It's strange.
I manually replaced the redundant newline with "^M".

But now it shows that "^M" and "^M" are different.

File: elpaca.el [hunk 5]
```diff
@@ -866,7 +866,7 @@ If it matches, the E associated with process has its STATUS updated."
          (timer   (process-get process :timer))
          (chunk   (concat parsed output))
          (lines   (split-string chunk "\n"))
-         (returnp (string-match-p "" chunk))
+         (returnp (string-match-p "^M" chunk))
```
Comment (lebensterben) [inline elpaca.el:869]: I've no idea what's going on here.

Comment (progfolio) [reply elpaca.el:869]: >  I manually replaced the redundant newline with "^M".

Won't work by typing "^" followed by "M". The original value was a control character for a return.
You'll have to type `C-q` followed by `RET`.

Comment (lebensterben) [reply elpaca.el]: It's fixed now.

**Comment by lebensterben** [comment]
This is a bug caused by parinfer-rust-mode.

**Comment by lebensterben** [comment]
Given that
- It confuses parinfer-rust-mode
- It confuses git-diff
- It appears as a string `"^M"` on github.com web interface but as a zero-width whitespace on github ios client.

Should we use "\n" instead?

**Comment by progfolio** [comment]
Merged! Thank you.

**Comment by progfolio** [comment]
> Given that
> 
>     * It confuses parinfer-rust-mode
> 
>     * It confuses git-diff
> 
>     * It appears as a string `"^M"` on github.com web interface but as a zero-width whitespace on github ios client.
> 
> 
> Should we use "\r" instead?

That should work. I've patched that as well.

**Pull request merged**